File: Data Structures & Algorithms/interleaving-string/submission-1.cpp
// class Solution {
//     int n1, n2, n3;
//     vector<vector<vector<bool>>> dp;
//     bool fun(int i, int j, int k, string &s1, string &s2, string &s3) {
//         if (i == n1 && j == n2 && k == n3) {
//             return true;
//         }

//         if(i < n1 && j < n2) {
//             if(s1[i] != s3[k] && s2[j] != s3[k]) return false;
//         }

//         if(i < n1) {
//             if(s1[i] == s3[k]) {
//                 if(fun(i+1, j, k+1, s1, s2, s3)) return true;
//             }
//         }

//         if(j < n2) {
//             if(s2[j] == s3[k]) {
//                 if(fun(i, j+1, k+1, s1, s2, s3)) return true;
//             } 
//         }

//         return false;
//     }
// public:
//     bool isInterleave(string s1, string s2, string s3) {
//         n1 = s1.length(); n2 = s2.length() ; n3 = s3.length();
//         if(n1 + n2 != n3) return false;
//         dp.assign(vector<)
//         return fun(0, 0, 0, s1, s2, s3);
//     }
// };


class Solution {
    int n1, n2, n3;
    vector<vector<int>> dp;

    bool fun(int i, int j, string &s1, string &s2, string &s3) {

        if(i == n1 && j == n2)
            return true;

        if(dp[i][j] != -1)
            return dp[i][j];

        int k = i + j;

        bool ans = false;

        if(i < n1 && s1[i] == s3[k]) {
            ans = ans || fun(i+1, j, s1, s2, s3);
        }

        if(j < n2 && s2[j] == s3[k]) {
            ans = ans || fun(i, j+1, s1, s2, s3);
        }

        return dp[i][j] = ans;
    }

public:
    bool isInterleave(string s1, string s2, string s3) {
        n1 = s1.length();
        n2 = s2.length();
        n3 = s3.length();

        if(n1 + n2 != n3) return false;

        dp.assign(n1+1, vector<int>(n2+1, -1));

        return fun(0,0,s1,s2,s3);
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null) return ans;
        Queue<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        while(!q.isEmpty()) {
            int sz = q.size();
            while(sz != 0) {
                TreeNode node = q.poll();
                if(node.left != null) q.offer(node.left);
                if(node.right != null) q.offer(node.right);
                if(sz == 1) ans.add(node.val);
                sz--;
            }
        }

        return ans;
    }
}
